### ISSUE TITLE
Importing foreign IGPUImage and IGPUBuffer objects & swapchain image creation

### ISSUE THREAD

**Issue by deprilula28** [Devsh-Graphics-Programming/Nabla#395]
## Description
We need to import foreign/external IGPUImage and IGPUBuffer objects for integration with Qt/UE/Unity/etc (and internal use with swapchain images), retiring [CVulkanForeignImage](https://github.com/Devsh-Graphics-Programming/Nabla/blob/master/src/nbl/video/CVulkanForeignImage.h).

We also need a solution for creation of swapchain images.

## Description of the related problem
Circular dependency between swapchain and its images: #246 

## Solution proposal
- Give `COpenGLImage` & `CVulkanImage` a special public constructor that takes the handle
- Don't destroy the foreign objects in the destructor
- Move over `COpenGLImage` & `CVulkanImage` to `include/nbl/video` (so they're in public headers for engine users)

**Comment by devshgraphicsprogramming** [comment]
>     * Swapchain needs to be in charge of:
>       
>       * Creating itself
>       * Creating foreign IGPUImages
>       * Presenting
> 
>     * Swapchain should have `std::atomic<bool> m_imageExists[]` to avoid creating multiple IGPUImages with the same VkImage

Not part of this issue.

**Comment by achalpandeyy** [comment]
## CUDA device vs Vulkan Physical Device
Match `VkPhysicalDeviceIDProperties::deviceUUID` with `cudaDeviceProp::uuid`.
```C
const bool match = memcmp((void *)vk_deviceUUID, cuda_deviceUUID, (size_t)VK_UUID_SIZE);
```

## Export from CUDA, import in Vulkan
Requires platform specific features.
```Cuda
#if defined(__linux__)
CUmemAllocationHandleType handleTypeFlag = CU_MEM_HANDLE_TYPE_POSIX_FILE_DESCRIPTOR; // file descriptors will be used
typedef int ShareableHandle;
#else
CUmemAllocationHandleType handleTypeFlag = CU_MEM_HANDLE_TYPE_WIN32; // NT HANDLEs will be used
typedef HANDLE ShareableHandle;
#endif
```
Windows-specific `LPSECURITYATTRIBUTES` is required when `CU_MEM_HANDLE_TYPE_WIN32` is used.
### How to export a buffer from CUDA
```C++
const size_t bufferSize = 1ull << 22; // some random size
CUmemGenericAllocationHandle cudaMemHandle;
{
    CUmemAllocationProp allocationProps = {};
    // fill up allocationProps
    allocationProps.requestHandleTypes = handleTypeFlag;

    cuMemCreate(&cudaMemHandle, bufferSize, &allocationProps, 0);
    cuMemMap(); // assign the memory handle obtained above to a VA space

    cuMemRelease(); // the actual freeing of memory allocation will happen on an eventual call to cuMemUnmap

    cuMemSetAccess(&accessDescriptor); // set access to the VA space
}

ShareableHandle shareableMemHandle;
cuMemExportToShareableHandle((void*)&shareableMemHandle, cudaMemHandle, handleTypeFlag, 0);
```
**Note:** Although we use some CUDA IPC APIs, no interprocess communication actually seems necessary.

After everything to cleanup:
- `cuMemUnmap`, `cuAddressFree`
- Close platform specific handles: `CloseHandle` on win32, something on Linux etc
### How to import a buffer in Vulkan

When creating a buffer, use `VkExternalMemoryBufferCreateInfo` in `pNext` of `VkBufferCreateInfo`, to specify the external memory handle types. For example `VK_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_WIN32_BIT` for win32 etc.

```C++
VkBufferCreateInfo bufferInfo = {};
bufferInfo.sType = VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO;
bufferInfo.size = size;
bufferInfo.usage = usage;
bufferInfo.sharingMode = VK_SHARING_MODE_EXCLUSIVE;

VkExternalMemoryBufferCreateInfo externalMemoryBufferInfo = {};
externalMemoryBufferInfo.sType = VK_STRUCTURE_TYPE_EXTERNAL_MEMORY_BUFFER_CREATE_INFO;
externalMemoryBufferInfo.handleTypes = VK_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_WIN32_BIT;
bufferInfo.pNext = &externalMemoryBufferInfo;
```

Importing memory into Vulkan is done by extending the `VkMemoryAllocateInfo` via its `pNext`.
For example, we use  `VkImportMemoryFdInfoKHR` to import memory from a POSIX handle and, `VkImportMemoryWin32HandleInfoKHR` to import from a Windows handle, which require the `VK_KHR_external_memory_fd` and `VK_KHR_external_memory_win32` respectively, both of these device extensions are **NOT** promoted to Vulkan 1.1.

Use `vkGetMemoryWin32HandlePropertiesKHR` to query to-be-imported memory's properties.

Importing creates a `VkDeviceMemory` object --the result of `vkAllocateMemory`, and must be freed with a `VkFreeMemory`
Also,
> Applications can import the same payload into multiple instances of Vulkan, into the same instance from which it was exported, and multiple times into a given Vulkan instance. In all cases, each import operation must create a distinct VkDeviceMemory object.

```C++
VkImportMemoryWin32HandleInfoKHR handleInfo = {};
handleInfo.sType = VK_STRUCTURE_TYPE_IMPORT_MEMORY_WIN32_HANDLE_INFO_KHR;
handleInfo.pNext = NULL;
handleInfo.handleType = VK_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_WIN32_BIT;
handleInfo.handle = handle;
handleInfo.name = NULL;

VkMemoryAllocateInfo memAllocation = {};
memAllocation.sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO;
memAllocation.pNext = (void *)&handleInfo;
..
```
### [WIP] Synchronization



## [WIP] Export from Vulkan, import in CUDA
Vulkan-CUDA interop is a CUDA 10 feature.

Allocate on Vulkan, import to CUDA, processing in CUDA

Memory allocations imported from Vulkan are represented in CUDA as a `cudaExternalMemory_t` instance.
Semaphores imported from Vulkan are represented in CUDA as a `cudaExternalSemaphore_t` instance.

`cudaImportExternalMemory` is used to import memory allocated by Vulkan.

Note: `VK_KHR_external_semaphore` and `VK_KHR_external_memory` are required but both are already promoted to Vulkan 1.1.

Get a windows `HANDLE` to export from Vulkan via `vkGetMemoryWin32HandleKHR`.

**Comment by devshgraphicsprogramming** [comment]
Fixed in #398 

@achalpandeyy open a separate PR or Issue about Semaphore and Device Export from Nabla
